Templates

johnbute opened the pull request:
### Motivation:

Currently, SwiftPM supports a handful of hardcoded templates that act as a starting point for user projects. These built-in templates, accessible via `swift package init`, allow users to initialize basic packages such as libraries, tools, executables, and macros. While `swift package init` remains valuable for initializing simple packages, many Swift developers face more complex requirements when starting a new project, whether that may be initializing an HTTPS server or creating a package that implements an OpenAPI specification. These more complex package initializations have to be achieved through third-party command-line tools or custom scripts, many of which operate outside the SwiftPM ecosystem. This disconnect increases the complexity and makes it harder for users to discover, share, or customize initialization workflows.

### Modifications:

Modifying Package.swift to store template metadata, added template utilities, new init function that supports old workflow and new template workflow, test template function, prompting system. Added show-templates function.
### Result:

Now, users can initialize packages based on templates from local disk, git, or registry. Furthermore, they can author templates and test them with test template. For more, read documentation.

Commit:
created command for templates + added commands for local and git

Commit:
initial push for templates in package.swift prototype

Commit:
fixtures + cleanup

Commit:
deleting unnecessary fixtures

Commit:
added new dependencies to handle addDependency functionality of templates

Commit:
build plan for building products and targets of type templates, will need to revisit in future

Commit:
fulfilling cases for addproduct, will need to revisit to make sure adding product of type template works

Commit:
fulfilling cases for addtarget, will need to revisit to make sure adding target of type template works

Commit:
show templates in a local package

Commit:
fullfilling switch case for snippets

Commit:
registering the show-templates command and adding more to init

Commit:
removing testing library options, to revisit and discuss

Commit:
added context switcher for swiftcommandstate

Commit:
make templates recognizable by package.swift

Commit:
dependency graph for swift targets and templates need to be revisited

Commit:
diagnostics added for template targets + products, will need to add tests for this too

Commit:
manifest loading and json parsing

Commit:
templates directory check added, as template executables should not be in sources

Commit:
added templates as a type of module, will need to revisit to edit behavior as an executable

Commit:
added templates as a type of module, will need to revisit to edit behavior as an executable pt2

Commit:
changed tools version when generating a new project, will need to revert back to major, minor once demo completed

Commit:
target and product descriptions syntax

Commit:
build parameters for templates, will need to revisit

Commit:
plugins support for executing templates

Commit:
buildsupport for templates, will need to revisit

Commit:
template creation workhorse

Commit:
build support for xcbuild

Commit:
reverting plugin test to normal

Commit:
printing out --experimental-dump-help

Commit:
capturing --experimental-dump-help and parsing it

cmcgee1024 committed:
Move new package description API to the array types and update version to 999.0.0

cmcgee1024 committed:
Fix show-executables subcommand so that it does not show templates

Fix show-templates subcommand so that it does not show snippets

Make the swift-package binary development environment independent

cmcgee1024 committed:
Use in-process API to run experimental dump help and decode JSON

Commit:
fixtures update

Commit:
Updating package.swift to include Argumnetparser dependency for workspace

Commit:
argument parsing dump-help

Commit:
running the executable

Commit:
fixed syntax issues + added functionality for arguments with enums

cmcgee1024 committed:
Remove template target and product types and use the template init options instead

cmcgee1024 committed:
Remove template target and product types and use the template init options instead

cmcgee1024 committed:
Match whitespace and formatting to original main branch

cmcgee1024 committed:
Match whitespace and formatting to original main branch

Commit:
formatting + starting to add git capabilities

Commit:
formatting + starting to add git capabilities

Commit:
Merge pull request #23 from christie-mcgee/reduce_target_product_types

Reduce changes compared to SwiftPM main branch

Commit:
git support + dependency mapping based on type

Commit:
git support + dependency mapping based on type

Commit:
Merge branch 'inbetween' of github.pie.apple.com:jbute/swift-pm-template into inbetween

Commit:
added requirements + ability to checkout a branch without creating new one

Commit:
added requirements + ability to checkout a branch without creating new one

Commit:
added git functionality to show-templates

Commit:
added git functionality to show-templates

Commit:
refactoring and organizing code

Commit:
refactoring and organizing code

Commit:
organizing code to differentiate between initializing regular package and initializing template package

Commit:
organizing code to differentiate between initializing regular package and initializing template package

Commit:
added documentation + reorganized show-templates code

Commit:
added documentation + reorganized show-templates code

Commit:
changed naming from template types (git, local, registry) to template source

Commit:
changed naming from template types (git, local, registry) to template source

Commit:
generating templates from package registry

Commit:
generating templates from package registry

Commit:
added registry support for show-templates

Commit:
added registry support for show-templates

Commit:
formatting + documentation + quality

Commit:
formatting + documentation + quality

cmcgee1024 committed:
Add instructions to the SwiftPM readme for trying out templates

cmcgee1024 committed:
Add instructions to the SwiftPM readme for trying out templates

cmcgee1024 committed:
Make the template option optional when initializing with a template

cmcgee1024 committed:
Make the template option optional when initializing with a template

cmcgee1024 committed:
On failure due to multiple available templates provide the list as part of the error message

cmcgee1024 committed:
On failure due to multiple available templates provide the list as part of the error message

Commit:
Merge pull request #36 from christie-mcgee/make_template_name_optional

Make template name optional

Commit:
Merge pull request #35 from christie-mcgee/readme_instructions

Add instructions to the SwiftPM readme for trying out templates

Commit:
ergonomics, reducing repetition of template keyword

Commit:
ergonomics, reducing repetition of template keyword

Commit:
added documentation + predetermined args

Commit:
added documentation + predetermined args

Commit:
changed args option to instead take any arguments after as predetermined args for templates

Commit:
changed args option to instead take any arguments after as predetermined args for templates

Commit:
inferring source of template

Commit:
inferring source of template

Commit:
fixing error regarding passing args to template

Commit:
fixing error regarding passing args to template

Commit:
args checking + file existing checks if local template option picked

Commit:
args checking + file existing checks if local template option picked

Commit:
made description, permissions, and initial type top level

Commit:
made description, permissions, and initial type top level

Commit:
made scaffolding top level command

Commit:
made scaffolding top level command

Commit:
added ability to view description of templates to show-templates

Commit:
added ability to view description of templates to show-templates

Commit:
new command swift package scaffold

Commit:
new command swift package scaffold

Commit:
added more clarity to .template parameter initialType

Commit:
added more clarity to .template parameter initialType

Commit:
registry stuff, might revert later, but for now we can keep

Commit:
registry stuff, might revert later, but for now we can keep

Commit:
added back to init templates

Commit:
added back to init templates

Commit:
prompting improvements

Commit:
prompting improvements

Commit:
added validate package, first-class testing support, temporary packages

Commit:
added validate package, first-class testing support, temporary packages

Commit:
added testing logs + revamped testing system for templates

Commit:
added testing logs + revamped testing system for templates

Commit:
reverted package-metadata changes

Commit:
reverted package-metadata changes

Commit:
fixed dummyRepositorymanager to conform to protocol

Commit:
fixed dummyRepositorymanager to conform to protocol

Commit:
possibility to edit format of test result to json too

Commit:
possibility to edit format of test result to json too

Commit:
bug fix where subcommand was prompted for when there was no valid subcommand to choose from

Commit:
bug fix where subcommand was prompted for when there was no valid subcommand to choose from

Commit:
run package clean after generating package

Commit:
run package clean after generating package

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring + error handling

Commit:
refactoring + error handling

Commit:
refactoring templates + fixing bug where git and template id was not registered + readding resolution to template if only one template in package

Commit:
refactoring templates + fixing bug where git and template id was not registered + readding resolution to template if only one template in package

Commit:
refactoring test template + changed structure holding prompting answers

Commit:
refactoring test template + changed structure holding prompting answers

Commit:
refactored directory manager to encapsulate shared logic between init and test

Commit:
refactored directory manager to encapsulate shared logic between init and test

Commit:
refactored TemplateManager based on shared logic between init + test

Commit:
refactored TemplateManager based on shared logic between init + test

3405691582 committed:
Add bootstrap option to install without building.

This allows the build action to be decoupled from the install action, so
packaging systems that expect to be able to do discrete installation
steps post-build can do so cheaply, without having to effectively start
the build over from scratch.

This is only intended solely for that use-case, so this is not wired up
to other part of the swiftpm build process and defaults to being
disabled.

jakepetroules committed:
Emit "Found unhandled resource" to observability scope

Libraries should never call print. Emit this diagnostic message to the observability scope instead.

Commit:
added end-to-end test for testing initializing a package from local template + fixed logic with handling of directories

Commit:
added end-to-end test for testing initializing a package from local template + fixed logic with handling of directories

Commit:
reducing PR size

Commit:
reducing PR size

Commit:
Merge pull request #46 from jbute/my-update

My update

Commit:
added test coverage to showExecutables to make sure template does not appear in output

Commit:
added test coverage to showExecutables to make sure template does not appear in output

xedin committed:
[Commands] Adopt changes to package manifest refactoring actions

Previously package manifest refactoring actions conformed to a custom
refactoring provider and produced a `PackageEdit` that, in addition to
manifest file changes, included a list of auxiliary files. This is no
longer the case and all package manifest refactoring actions are now
responsible only for manifest file transformations, the rest is handled
by the commands themselves.

bripeticca committed:
Fix for .when(traits:) condition not working for multiple traits (#9015)

The culprit here is the fact that we were essentially doing an AND
boolean check across all the traits in the target dependency condition,
when we should be doing an OR boolean check.

If at least one trait is enabled in the list of traits in the target
dependency condition, then the target dependecy should be considered
enabled as well.

Commit:
added more tests, updated showtemplates test

Commit:
added more tests, updated showtemplates test

JeffESchmitz committed:
Fix: Generate tests for executable/tool with `--enable-swift-testing` (#9032)

Updated `InitPackage` to handle `.executable` and `.tool` types when
generating test files with SwiftTesting. Added corresponding tests to
verify correct test file content for these package types.

### Motivation:

The `swift package init` command did not generate a `Tests` directory
for executable or tool packages when the `--enable-swift-testing` flag
was used. This was caused by logic in
`Sources/Workspace/InitPackage.swift` that incorrectly skipped test
generation for these package types. This change fixes that behavior,
ensuring that tests are created as expected.

### Modifications:

- In `Sources/Workspace/InitPackage.swift`, the `writeTests()` function
was modified to proceed with test generation for `.executable` and
`.tool` package types.
- In the same file, the `writeTestFileStubs()` function was updated to
include `.executable` and `.tool` types, allowing the creation of
standard library-style test files for them.

### Result:

After this change, running `swift package init --type executable
--enable-swift-testing` or `swift package init --type tool
--enable-swift-testing` correctly generates a `Tests` directory with
sample test files, aligning the tool's behavior with its intended
functionality.

### Reproduction and Verification:

#### Reproducing the Issue (on `main` branch):

1. Create a temporary directory: `mkdir -p /tmp/test-repro && cd
/tmp/test-repro`
2. Run `swift package init --type executable --enable-swift-testing`.
3. Verify that no `Tests` directory is created by running `ls -F`.

#### Verifying the Fix (on this branch):

1. Build the package manager: `swift build`
2. Find the built executable: `find .build -name swift-package`
3. Create a temporary directory: `mkdir -p /tmp/test-fix && cd
/tmp/test-fix`
4. Run the newly built package manager to init a project:
`../.build/arm64-apple-macosx/debug/swift-package init --type executable
--enable-swift-testing` (adjust path as needed).
5. Verify that a `Tests` directory is now created by running `ls -F`.

Commit:
added more test coverage

Commit:
added more test coverage

Commit:
added observability scope to templatepathresolver + ssh error message for special case where authentication is needed

Commit:
added observability scope to templatepathresolver + ssh error message for special case where authentication is needed

heckj committed:
adds traits to dependency output in show-dependencies --json and --text (#9034)

Adds traits to the output of the show-dependencies command for `--text`
and `--json` output.

### Motivation:

Exposes the default traits, if any, that are enabled by the dependencies
a package.

Resolves: #9033 

### Modifications:

Extended the logic in show-dependencies to add in the `enabledTraits`
for a resolved package.

### Result:

The output of show-dependencies includes output akin to `(traits:
FirstTrait, SecondTrait)` is the resolved package has traits and
includes the traits that it has. If the package doesn't offer traits, no
additional/new content is displayed in `--text`.
The JSON provided from `show-dependencies --json` includes a `traits`
property, empty if there aren't any traits enabled.

Commit:
solved bug where test templates would not work as current directory was erroneous if ran in different context

Commit:
solved bug where test templates would not work as current directory was erroneous if ran in different context

bkhouri committed:
Tests: Migrate 2 suites to Swift Testing (#9017)

Migrate a few suites to Swift Teting, namely,
  - SwiftSDKCommandTests
  - MultiRootSupportTests

Depends on: #9012
Relates to: #8997
issue: rdar://157669245

Commit:
fixed bug where answers to arguments of a specific argument branch were not shared outside that branch + improved dry-run output to print command line arguments

Commit:
fixed bug where answers to arguments of a specific argument branch were not shared outside that branch + improved dry-run output to print command line arguments

bkhouri committed:
Tests: Migrate BuildSystemDelegateTests to Swift Testing and augment (#9013)

Migrate the `BuildSystemDelegateTests` test to Swift Testing and augment
the test to run against both the Native and SwiftBUild build system, in
addition to the `debug` and `release` build configuration.

Depends on: #9012
Relates to: #8997
issue: rdar://157669245

Commit:
added predefined arguments to testing, however will need to revisit to get clarity on handling of subcommands

Commit:
added predefined arguments to testing, however will need to revisit to get clarity on handling of subcommands

Commit:
added ability to test specific branches of decision tree, need to flesh out however

Commit:
added ability to test specific branches of decision tree, need to flesh out however

owenv committed:
Stage in fix for toolsetRunner test (#9086)

Ensure this test doesn't start failing when I merge
https://github.com/swiftlang/swift-build/pull/765

Commit:
changes to usability to resolve to latest registry version

Commit:
changes to usability to resolve to latest registry version

heckj committed:
initial, simple SDK documentation (#9004)

Provides some basic detail for Swift cross-compilation SDKs, referencing
the installation links on swift.org, the evolution proposal, and the SDK
generator code.

resolves #8864, to finalize the documentation of the CLI commands

heckj committed:
Traits docs (#9048)

Adds documentation for the Swift 6.1 traits feature

### Motivation:

- Show how to consume packages that provide traits
- Show how to present traits from your package(s)
- Updating PackageDescription API to use slightly more readable content
related to traits.

### Modifications:

- updated PackageDescription API around traits
- added curation (organization) for Traits and Package/Dependency/Traits
- extended `Using Dependencies` article to touch on traits and how to
consume packages that provide them
- added an article to share patterns to use when providing a traits from
your own package

### Result:

Updated content and one additional article in the central documentation
for Swift Package Manager that illustrates how to consume and provide
packages with traits.

---------

Co-authored-by: Franz Busch <privat@franz-busch.de>
Co-authored-by: Bri Peticca <bripeticca@gmail.com>

Commit:
enforcing templates to have a corresponding product and plugin

Commit:
enforcing templates to have a corresponding product and plugin

Commit:
error handling more gracefully, with reporting from observability scope

Commit:
error handling more gracefully, with reporting from observability scope

shahmishal committed:
Bump the Swift version to 6.3 (#9028)

Commit:
fixed permissions to query for them only once during process + refactored coordinator + plugin managers

Commit:
fixed permissions to query for them only once during process + refactored coordinator + plugin managers

owenv committed:
Drop unknown platforms when generating PIF (#9021)

Some existing packages use `.when(platforms: [.custom("DoesNotExist")])`
to unconditionally deactivate settings. Stop asserting on these unknown
platforms

bkhouri committed:
Test: Add issue to test (#9084)

Add an issue to a test so we can track it's withKnownIssue resolution

bkhouri committed:
Test: Migrate CFamilyTargetTests to Swift Testing and augment (#9014)

Migrate the `CFamilyTargetTests` test to Swift Testing and augment the
test to run against both the Native and SwiftBUild build system, in
addition to the `debug` and `release` build configuration.

Depends on: #9013
Relates to: #8997
issue: rdar://157669245

bkhouri committed:
Tests: Update helpers to support both test libraries (#9082)

The `_InternalTestSupport` and `_InternalBuildTestSupport` modules have
XCTest assert helper functions, that, when called from a Swift Testing
test, is essentially a no-op. During migration to Swift Testing, this
could lead to false positives if the helpers function are not converted
to use Swift Testing APIs.

During the transition, update all helper functions to detect whether the
caller is executing in the XCTest or Swift Testing context, and call the
appropriate API given the library.

Commit:
Merge pull request #48 from jbute/update

Update main branch

Commit:
refactored tests + added missing ones

Commit:
refactored + fixed small implementation details + tests for make dependencies, and filled some previous test gaps

Commit:
rebasing changes

Commit:
changed template running system to reflect @ParentCommand changes

Commit:
tests + revamped prompting systems

Commit:
formatting

Commit:
changes to reflect github comments

Commit:
formatting + reverting unnecessary changes

Commit:
documentation

Commit:
documentation for authoring, initializing from, and testing templates

Commit:
documentation improvements

Commit:
fixed optional bug when prompting + added templates examples

Commit:
fixed printing

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' of github.pie.apple.com:jbute/swift-pm-template

Commit:
fixed merge conflicts

Commit:
Merge pull request #51 from jbute/update

Update

Commit:
Merge branch 'main' into inbetween

Commit:
bumped version number

Commit:
merged with internal repo due to missing commits

Commit:
reverted swift-version

Commit:
reducing PR size by getting rid of unwanted changes

Commit:
reducing PR

Commit:
removed source conflicts in fixtures

Commit:
formatting

Commit:
removed redeclarations mistake during merge

cmcgee1024 commented:
@swift-ci please test

Commit:
added template modules to cmake

Commit:
fixed cmake script error, testTemplateComand tests started, edited function names for clarity, add test for compiling manifest file for templates, removed tests regarding building a template as they are already handled by E2E tests

johnbute commented:
@swift-ci

johnbute commented:
@swift-ci

Commit:
revamped parsing, prompting logic for templates, hid test template command until refactored changes

Commit:
merge conflicts between templates and traits

Commit:
fixed suite parrelilization issues with template

Commit:
resolving merge conflict

Commit:
redeclaration error fix

johnbute commented:
@swift-ci please test

Commit:
fixed tests and disabled directory management tests, to investigate as they pass locally but not in ci

Commit:
disabling template tests for working toolchain

Commit:
added git version resolution to latest when version requirement omitted

heckj committed:
shifting documenantation into PackageManagerDocs

johnbute committed:
Merging documentation changes

shifting documenantation into PackageManagerDocs

heckj commented on Examples/init-templates/Plugins/PartsServicePlugin/PartsServicePlugin.swift:
```diff
@@ -0,0 +1,24 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// A plugin that kickstarts the Parts Service executable.
```

heckj commented on Examples/init-templates/Plugins/ServerTemplatePlugin/ServerTemplatePlugin.swift:
```diff
@@ -0,0 +1,51 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// A plugin that kickstarts the Server Template executable.
```

heckj commented on Examples/init-templates/Plugins/Template1Plugin/Template1Plugin.swift:
```diff
@@ -0,0 +1,54 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// A plugin that kickstarts the Template1 executable.
```

heckj commented on Examples/init-templates/Tests/TemplateTest.swift:
```diff
@@ -0,0 +1,80 @@
+import Foundation
+import Testing
+
+// a possible look into how to test templates
```
```suggestion
// An example of how to test templates.
```

heckj commented on Examples/init-templates/Tests/TemplateTest.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    // function for running a process given arguments, executable, and a directory
```
```suggestion
    // Function for running a process given arguments, executable, and a directory
```

heckj commented on Examples/init-templates/README.md:
```diff
@@ -0,0 +1,21 @@
+
+---
+
+This template project is a simple example of how a template author can create a template and generate a swift projects, utilizing the `swift package init` capability in swift package manager (to come).
```
```suggestion
This project provides a simple example to show how to author a template to generate Swift projects, utilizing the template feature included within the `swift package init` command of the Swift Package Manager.
```

heckj commented on Examples/init-templates/README.md:
```diff
@@ -0,0 +1,21 @@
+
+## Parts Service
+
+The parts service template can generate a REST service using Hummingbird (app server), and Fluent (ORM) with configurable database management system (SQLite3, and PostgreSQL). There are various switches to customize your project.
```
```suggestion
The parts service template generates a REST service using Hummingbird (app server) and Fluent (ORM) with a configurable database management system (SQLite3 and PostgreSQL). The template includes various switches to customize your project.
```

heckj commented on Fixtures/Miscellaneous/InferPackageType/Plugins/initialTypeEmptyPlugin/main.swift:
```diff
@@ -0,0 +1,12 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// The plugin that kickstarts the template executable.
```

heckj commented on Fixtures/Miscellaneous/InferPackageType/Plugins/initialTypeMacroPlugin/main.swift:
```diff
@@ -0,0 +1,12 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// The plugin that kickstarts the template executable.
```

heckj commented on Fixtures/Miscellaneous/InferPackageType/Plugins/initialTypeToolPlugin/main.swift:
```diff
@@ -0,0 +1,12 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable
```
```suggestion
/// The plugin that kickstarts the template executable.
```

heckj commented on Fixtures/Miscellaneous/ShowTemplates/app/Plugins/GenerateFromTemplatePlugin/plugin.swift:
```diff
@@ -0,0 +1,29 @@
+
+import PackagePlugin
+
+/// plugin that will kickstart the template executable=≠≠
```
```suggestion
/// The plugin that kickstarts the template executable.
```

heckj commented on Sources/Commands/SwiftBuildCommand.swift:
```diff
@@ -103,13 +103,14 @@ struct BuildCommandOptions: ParsableArguments {
     @Option(help: "Build the specified product.")
     var product: String?
 
+    /*
```
should this continue to be commented out, or removed?

heckj commented on Sources/CoreCommands/SwiftCommandState.swift:
```diff
@@ -1374,3 +1374,74 @@ extension Basics.Diagnostic {
 }
 
+extension SwiftCommandState {
+    /// Temporarily switches to a different package directory and executes the provided closure.
```
```suggestion
    /// Temporarily switch to a different package directory and execute the closure you provide.
```

heckj commented on Sources/CoreCommands/SwiftCommandState.swift:
```diff
@@ -1374,3 +1374,74 @@ extension Basics.Diagnostic {
+    /// including workspace initialization, file system changes, and cleanup.
+    ///
+    /// - Parameters:
+    ///   - packagePath: The absolute path to switch to
```
```suggestion
    ///   - packagePath: The absolute path to switch to.
```

heckj commented on Sources/CoreCommands/SwiftCommandState.swift:
```diff
@@ -1374,3 +1374,74 @@ extension Basics.Diagnostic {
+    ///
+    /// - Parameters:
+    ///   - packagePath: The absolute path to switch to
+    ///   - createPackagePath: Whether to create the directory if it doesn't exist
```
```suggestion
    ///   - createPackagePath: A Boolean value that indicates whether to create the directory if it doesn't exist.
```

heckj commented on Sources/CoreCommands/SwiftCommandState.swift:
```diff
@@ -1374,3 +1374,74 @@ extension Basics.Diagnostic {
+    /// - Parameters:
+    ///   - packagePath: The absolute path to switch to
+    ///   - createPackagePath: Whether to create the directory if it doesn't exist
+    ///   - perform: The closure to execute in the temporary workspace context
```
```suggestion
    ///   - perform: The closure to execute in the temporary workspace context.
```

heckj commented on Sources/CoreCommands/SwiftCommandState.swift:
```diff
@@ -1374,3 +1374,74 @@ extension Basics.Diagnostic {
+    ///   - createPackagePath: Whether to create the directory if it doesn't exist
+    ///   - perform: The closure to execute in the temporary workspace context
+    /// - Returns: The result of the performed closure
+    /// - Throws: Any error thrown by the closure or during workspace setup
```
```suggestion
    /// - Returns: The result of the performed closure.
    /// - Throws: Any error thrown by the closure or during workspace setup.
```

heckj commented on Sources/Runtimes/PackageDescription/Target.swift:
```diff
@@ -229,6 +229,23 @@ public final class Target {
 
+    public var templateInitializationOptions: TemplateInitializationOptions?
+
+    public enum TemplateType: String {
```
```suggestion
    /// The kind of Swift project that the template generates.
    public enum TemplateType: String {
```

heckj commented on Sources/Runtimes/PackageDescription/Target.swift:
```diff
@@ -229,6 +229,23 @@ public final class Target {
         case plugin(name: String, package: String?)
     }
 
+    public var templateInitializationOptions: TemplateInitializationOptions?
```
```suggestion
    /// Options for how a template generates a Swift project.
    public var templateInitializationOptions: TemplateInitializationOptions?
```

heckj commented on Sources/Runtimes/PackageDescription/Target.swift:
```diff
@@ -1236,6 +1270,89 @@ public final class Target {
 }
 
+public extension [Target] {
+    @available(_PackageDescription, introduced: 6.3.0)
```
```suggestion
    /// The declaration of a Swift package template.
    @available(_PackageDescription, introduced: 6.3.0)
```

(I'm not certain this is accurate - I wasn't clear when I cobbled this what the declaration of `template` within a list of targets was meant to convey, and where it's used.)

heckj commented on Sources/Runtimes/PackageDescription/Target.swift:
```diff
@@ -1562,3 +1679,48 @@ extension Target.PluginUsage: ExpressibleByStringLiteral {
+/// Supported types are ``allowNetworkConnections(scope:reason:)`` and ``writeToPackageDirectory(reason:)``.
+@available(_PackageDescription, introduced: 6.3.0)
+public enum TemplatePermissions {
+    /// Create a permission to make network connections.
```
```suggestion
    /// The permission to make network connections.
```

heckj commented on Sources/Runtimes/PackageDescription/Target.swift:
```diff
@@ -1562,3 +1679,48 @@ extension Target.PluginUsage: ExpressibleByStringLiteral {
+    /// The command plug-in requires permission to make network connections. The `reason` string is shown
+    /// to the user at the time of request for approval, explaining why the plug-in is requesting access.
+    ///   - Parameter scope: The scope of the permission.
+    ///   - Parameter reason: A reason why the permission is needed. This is shown to the user when permission is sought.
```
```suggestion
    /// The plug-in requires permission to make network connections. Swift Package Manager shows the `reason` string
    /// to the user at the time of request for their approval, explaining why the plug-in is requesting access.
    ///   - Parameter scope: The scope of the permission.
    ///   - Parameter reason: A reason why the permission is needed; shown to the user when permission is sought.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/CreatingSwiftPackage.md:
```diff
@@ -81,3 +81,78 @@ This creates a package with:
+
+### Creating a Package based on a custom template
+
+Swift Package Manager can create packages based on custom templates taht authors distribute as Swift packages.
```
```suggestion
Swift Package Manager can create packages based on custom templates that authors distribute as Swift packages.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/CreatingSwiftPackage.md:
````diff
@@ -81,3 +81,78 @@ This creates a package with:
+$ swift package init --type MyTemplate --path /path/to/template
+```
+
+The template prompts you for configuration options during initialization:
````
```suggestion
If there are options, the template prompts you for configuration during initialization:
```

(is this correct? I assumed that it's possible to create a template without options)

heckj replied:
```suggestion
The template prompts you for the package configuration options during initialization:
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/CreatingSwiftPackage.md:
````diff
@@ -81,3 +81,78 @@ This creates a package with:
+$ swift package init --type MyTemplate --url https://github.com/author/template --revision abc123
+```
+
+You can provide template arguments directly to skip interactive prompts:
````
```suggestion
You may provide template arguments proactively in order to skip the interactive prompts:
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+## Overview
+
+Custom _templates_ allow generation of packages whose functionality goes beyond the hard-coded templates provided by Swift Package Manager. Package templates are written in Swift using Swift Argument Parser and the `PackagePlugin` API provided by the Swift Package Manager.
```
```suggestion
Custom _templates_ allow Swift Package Manager to generate of packages whose functionality goes beyond the hard-coded templates. Package templates are written in Swift using Swift Argument Parser to provide arguments for generating a Swift package.
```

heckj reviewed (commented):
I'm having to submit this as multiple rounds of comments with suggestions for updated wording in documentation (and some questions inline). I'm exceeding what GitHub's review "javascript app" can handle in a browser with the size of this draft PR - but hopefully these are all easy to accommodate or understand what I'm suggesting.

heckj reviewed (commented):
I'm having to submit this as multiple rounds of comments with suggestions for updated wording in documentation (and some questions inline). I'm exceeding what GitHub's review "javascript app" can handle in a browser with the size of this draft PR - but hopefully these are all easy to accommodate or understand what I'm suggesting.

heckj commented on Sources/PackageModel/Manifest/TargetDescription.swift:
```diff
@@ -194,6 +194,45 @@ public struct TargetDescription: Hashable, Encodable, Sendable {
+        case empty
+    }
+
+    public enum TemplateNetworkPermissionScope: Hashable, Codable, Sendable {
```
```suggestion
    /// Network permissions that a template may require.
    public enum TemplateNetworkPermissionScope: Hashable, Codable, Sendable {
```

heckj commented on Sources/PackageModel/Manifest/TargetDescription.swift:
```diff
@@ -194,6 +194,45 @@ public struct TargetDescription: Hashable, Encodable, Sendable {
+        case docker
+        case unixDomainSocket
+
+        public init?(_ scopeString: String, ports: [Int]) {
```
```suggestion
        /// Create a network permission with the scope and ports that you provide.
        public init?(_ scopeString: String, ports: [Int]) {
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/CreatingSwiftPackage.md:
````diff
@@ -81,3 +81,78 @@ This creates a package with:
+Build of product 'ServerTemplate' complete! (1.1s)
+```
+
+Templates support the same versioning options as regular Swift package dependencies:
````
```suggestion
Templates support the same versioning options as other Swift package dependencies:
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+#  Creating Package Templates
+
+This guide provides a brief overview of Swift Package Manager templates, describes how a package can make use of templates, and shows how to get started writing your own templates.
```
```suggestion
Create a template plugin for Swift Package Manager to generate custom Swift packages based on the custom inputs that you define.
```

^ The first sentence in the article is called an "Abstract" and we generally want it to be a short, imperative sentence that tells the reader what they'll be learning if they jump in and read it. The abstract is presented alongside the title to the create a sort of description of what they'll learn and what core problem its solving.

I suspect there could easily be better title and abstract choices here, but wanted to at least suggest a placeholder that was closer to what we should be aiming for.

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+Custom _templates_ allow generation of packages whose functionality goes beyond the hard-coded templates provided by Swift Package Manager. Package templates are written in Swift using Swift Argument Parser and the `PackagePlugin` API provided by the Swift Package Manager.
+
+A template is represented in the SwiftPM package manifest as a target of the `templateTarget` type and should be available to other packages by declaring a corresponding `template` product. Source code for a template is normally located in a directory under the `Templates` directory in the package, but this can be customized. However, as seen below, authors will also need to write the source code ...
```
```suggestion
Swift Package Manager represents a template in the package manifest as a target of the `templateTarget` type.
Make the template available to other packages by declaring a corresponding `template` product.
Source code for a template is normally located in a directory under the `Templates` directory in the package, but you can customize its location.
The template also requires a template plugin to generate the content for the Swift package.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+A template is represented in the SwiftPM package manifest as a target of the `templateTarget` type and should be available to other packages by declaring a corresponding `template` product. Source code for a template is normally located in a directory under the `Templates` directory in the package, but this can be customized. However, as seen below, authors will also need to write the source code ...
+
+Templates are an abstraction of two types of modules:
+- a template _executable_ that performs the file generation and project setup
```
```suggestion
- a template _executable_ that generates the file generation and provides the arguments for the template.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+### Declaring a template in the package manifest
+
+Like all package components, templates are declared in the package manifest. This is done using a `templateTarget` entry in the `targets` section of the package. Templates must be visible to other packages in order to be ran. Thus, there needs to be a corresponding `template` entry in the `products` section as well:
```
```suggestion
Like all package components, declare templates in the package manifest. Use the `templateTarget` entry in the `targets` section of the package. Templates must be visible to other packages in order to run. Thus, there needs to be a corresponding `template` entry in the `products` section as well:
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
````diff
@@ -0,0 +1,458 @@
+Like all package components, templates are declared in the package manifest. This is done using a `templateTarget` entry in the `targets` section of the package. Templates must be visible to other packages in order to be ran. Thus, there needs to be a corresponding `template` entry in the `products` section as well:
+
+```swift
+// swift-tools-version: 6.1
````
```suggestion
// swift-tools-version: 6.3
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+The `templateTarget` declares the name and capability of the template, along with its dependencies. The `initialPackageType` specifies the base package structure that SwiftPM will set up before invoking the template — this can be `.library`, `.executable`, `.tool`, `.buildToolPlugin`, `.commandPlugin`, `.macro`, or `.empty`.
+
+The Swift script files that implement the logic of the template are expected to be in a directory named the same as the template, located under the `Templates` subdirectory of the package. The template also expects Swift script files in a directory with the same name as the template, alongside a `Plugin` suffix, located under the `Plugins` subdirectory of the package.
```
```suggestion
The Swift Package Manager expects the Swift script files that implement the logic of the template to be in a directory with the same as the template, located within a `Templates` subdirectory of the package. The Package Manager also expects Swift script files in a directory with the same name as the template, alongside a `Plugin` suffix, located under the `Plugins` subdirectory of the package.
```

This is complex enough that it's a bit hard to parse right off the bat. I'd suggest that we include an example directory structure inline here - such as the output of the `tree` command, illustrating the expectations to make it easier to understand.

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+The Swift script files that implement the logic of the template are expected to be in a directory named the same as the template, located under the `Templates` subdirectory of the package. The template also expects Swift script files in a directory with the same name as the template, alongside a `Plugin` suffix, located under the `Plugins` subdirectory of the package.
+
+The `template` product is what makes the template visible to other packages. The name of the template product must match the name of the target.
```
```suggestion
Declare the `template` product to make the template visible to other packages. The name of the template product must match the name of the target.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+#### Template target dependencies
+
+The dependencies specify the packages that will be available for use by the template executable. Each dependency can be any package product. Commonly this includes Swift Argument Parser for command-line interface handling, but can also include utilities for file generation, string processing, or network requests if needed.
```
```suggestion
The dependencies specify the packages that available for use by the template executable. Each dependency can be any package product. A common example is the Swift Argument Parser, which provides arguments, options, and flags for command-line interface handling, but can also include utilities for file generation, string processing, or network requests.
```

heckj commented on Sources/PackageManagerDocs/Documentation.docc/Templates.md:
```diff
@@ -0,0 +1,458 @@
+
+## Testing Templates
+
+SwiftPM provides a built-in command for testing templates comprehensively:
```
```suggestion
Swift Package Manager provides a built-in command for testing templates:
```

heckj reviewed (commented):
core docs review and wording suggestions

heckj reviewed (commented):
core docs review and wording suggestions

johnbute committed:
Update Examples/init-templates/Plugins/PartsServicePlugin/PartsServicePlugin.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Examples/init-templates/Plugins/ServerTemplatePlugin/ServerTemplatePlugin.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Examples/init-templates/Plugins/Template1Plugin/Template1Plugin.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Examples/init-templates/Plugins/Template2Plugin/Template2Plugin.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Examples/init-templates/Templates/Template1/Template.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageManagerDocs/Documentation.docc/CreatingSwiftPackage.md

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageManagerDocs/Documentation.docc/Templates.md

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageManagerDocs/Documentation.docc/Templates.md

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update README.md

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageModel/Manifest/TargetDescription.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageModel/Manifest/TargetDescription.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageModel/Manifest/TargetDescription.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageModel/Manifest/TargetDescription.swift

Co-authored-by: Joseph Heck <j_heck@apple.com>

johnbute committed:
Update Sources/PackageManagerDocs/Documentation.docc/Templates.md

Co-authored-by: Joseph Heck <j_heck@apple.com>

Commit:
fixed tests so that they can run in parallel (or serialized) without process pollution

Commit:
documentation change + reverted InitPackage formatting

Commit:
fixed bug in tests and serialized templates

Commit:
fixed formatting issue in tests + refactored testTmplate

Commit:
updated documentation based on joe's suggestions

Commit:
fixing test crash where force-unwrapping optional

Commit:
disabling test due to fatal error crash

johnbute commented:
@swift-ci please test

Commit:
refactored completly the test template argument parsing + logic

Commit:
fixed cmake files + removed deprecated options (not needed by vscode anymore

Commit:
Merge branch 'main' into templates

Commit:
disabling directory tests

johnbute commented:
@swift-ci please test

Commit:
checking if TemplatePath resolution is the issue

johnbute commented:
@swift-ci please test

Commit:
disabling Templateresoluttiontopackageworkflow test for ci

johnbute commented:
@swift-ci please test

Commit:
reverted disabling of tests, as issue is not scoped to a specific test

johnbute commented:
@swift-ci please test

cmcgee1024 committed:
Merge branch 'main' of https://github.com/swiftlang/swift-package-manager into templates

cmcgee1024 committed:
Add missing workers argument to plugin invoke

cmcgee1024 committed:
Add support for dumping help from the swift package init workflow

When invoking `swift package init --url ...` it should be possible
to extract the tool information for prompting by an external
application by adding `-- --dump-help` at the end. The application
can then prompt the user and re-run it with all of the choices
after the `--` delimiter to generate the template.

cmcgee1024 committed:
Merge branch 'main' of https://github.com/swiftlang/swift-package-manager into templates

cmcgee1024 committed:
Remove initial template type as an option for templates

Templates will use an empty, but functional package manifest, which
has fewer surfaces that can change and break templates. Remove the
option from the package manifest for template declarations.

Update the examples and documentation to reflect these changes.

johnbute committed:
Merge branch 'main' into pr-9773

johnbute committed:
merge with main

johnbute committed:
Merge remote-tracking branch 'origin/main' into templates

johnbute committed:
Merge branch 'templates' of https://github.com/johnbute/fork-swift-package-manager into templates

johnbute committed:
fixed bug while merging
